android/src/com/mapswithme/maps/widget/placepage/PlaceOpeningHoursAdapter.java

```diff
@@ -17,46 +19,84 @@
 
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Collections;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
 
 public class PlaceOpeningHoursAdapter extends RecyclerView.Adapter<PlaceOpeningHoursAdapter.ViewHolder>
 {
-  private Timetable[] mTimetables = {};
-  private int[] closedDays = null;
+  private List<WeekScheduleData> mWeekSchedule = Collections.emptyList();
 
   public PlaceOpeningHoursAdapter() {}
 
-  public PlaceOpeningHoursAdapter(Timetable[] timetables)
+  public PlaceOpeningHoursAdapter(Timetable[] timetables, int firstDayOfWeek)
   {
-    setTimetables(timetables);
+    setTimetables(timetables, firstDayOfWeek);
   }
 
-  public void setTimetables(Timetable[] timetables)
+  public void setTimetables(Timetable[] timetables, int firstDayOfWeek)
   {
-    mTimetables = timetables;
-    closedDays = findUnhandledDays(timetables);
-    notifyDataSetChanged();
-  }
+    int[] weekDays = null;
+    if (firstDayOfWeek == Calendar.SUNDAY)
+      weekDays = new int[]{1, 2, 3, 4, 5, 6, 7};
+    else
+      weekDays = new int[]{2, 3, 4, 5, 6, 7, 1};
 
-  private int[] findUnhandledDays(Timetable[] timetables)
-  {
-    List<Integer> unhandledDays = new ArrayList<>(Arrays.asList(1, 2, 3, 4, 5, 6, 7));
+    final List<WeekScheduleData> scheduleData = new ArrayList<>();
 
-    for (Timetable tt : timetables) {
-      for (int weekDay : tt.weekdays) {
-        unhandledDays.remove(Integer.valueOf(weekDay));
+    // timetables array contains only working days. We need to fill non working gaps
+    for (int idx=0; idx < weekDays.length; idx++)
+    {
+      int weekDay = weekDays[idx];
+
+      Timetable tt = findScheduleForWeekDay(timetables, weekDay);
+      if (tt != null)
+      {
+        int startWeekDay = weekDays[idx];
+        while(idx < weekDays.length && tt.containsWeekday(weekDays[idx]))
+          idx++;
```

Comment:
This implementation looks complex, is it possible to simplify it?

Reply:
I'll think about it. The idea is that we split array of days: `{1,2,3,4,5,6,7}` or `{2,3,4,5,6,7,1}` into ranges:

* [d1-d2] -> timetable1,
* [d3-d4] -> null,
* [d5-d6] -> timetable2,
* [d7-d8] -> timetable1, 
* etc. 

So I use `idx` to find first and last day of a range.

I created unittests to make sure implementation could handle complex cases. If something is not parsed correctly we will update the test 😉